staking-pool/src/lib.rs

```diff
@@ -90,358 +109,513 @@ impl Default for StakingContract {
-                self.epoch_infos.push(epoch_info);
-            }
-        }
-
-        // The total stake can also be 0 if there were no staking.
-        if total_stake > 0 {
-            let total_stake = U256::from(total_stake);
-            for (account_id, stake) in rewarded_accounts {
-                let staked_reward =
-                    (total_staked_reward * U256::from(stake) / total_stake).as_u128();
-                let unstaked_reward =
-                    (total_unstaked_reward * U256::from(stake) / total_stake).as_u128();
-                {
-                    let account = self.active_accounts.get_mut(&account_id).unwrap();
-                    account.staked += staked_reward;
-                    account.unstaked += unstaked_reward;
-                }
-                if let Some(stake) = new_epoch_info.stakes.get_mut(&account_id) {
-                    *stake += staked_reward;
-                    self.total_stake += staked_reward;
-                }
-            }
+        if self.internal_ping() {
+            self.restake();
         }
-
-        self.epoch_infos.push(new_epoch_info);
-
-        // Moving inactive accounts towards archiving.
-        for (account_id, account) in std::mem::take(&mut self.active_accounts) {
-            if self
-                .epoch_infos
-                .iter()
-                .any(|epoch_info| epoch_info.stakes.contains_key(&account_id))
-            {
-                self.active_accounts.insert(account_id, account);
-            } else {
-                self.inactive_accounts.insert(account_id, account);
-            }
-        }
-
-        self.last_locked_account_balance = env::account_locked_balance();
-        self.last_account_balance = env::account_balance();
     }
 
-    /// Call to deposit money.
+    /// Deposits the attached amount into the inner account of the predecessor.
     #[payable]
     pub fn deposit(&mut self) {
-        self.ping();
+        let need_to_restake = self.internal_ping();
+
         let account_id = env::predecessor_account_id();
-        let mut wrapped_account = self.pull_account(account_id);
-        wrapped_account.account.unstaked += env::attached_deposit();
-        self.save_account(wrapped_account);
+        let mut account = self.get_account(&account_id);
+        let amount = env::attached_deposit();
+        account.unstaked += amount;
+        self.save_account(&account_id, &account);
+        self.last_total_balance += amount;
+
+        env::log(
+            format!(
+                "@{} deposited {}. New unstaked balance is {}",
+                account_id, amount, account.unstaked
+            )
+            .as_bytes(),
+        );
+
+        if need_to_restake {
+            self.restake();
+        }
     }
 
     /// Withdraws the non staked balance for given account.
+    /// It's only allowed if the `unstake` action was not performed in the recent 3 epochs.
     pub fn withdraw(&mut self, amount: U128) {
```

Comment:
When an account has withdraw all of their tokens, shouldn't it be deleted from the map?